### PR TITLE
feat: add flake8-bugbear

### PR DESCRIPTION
**Pull request by jnoortheen**
## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

### CODE REVIEW

**Comment by jnoortheen** [comment]
>  the question is whether we want to introduce yet another linter.


This contains some important bug detection IMO

File: amalgamate.py [hunk 1]
```diff
@@ -362,7 +362,7 @@ def format_lazy_import(names):
 def format_from_import(names):
     """Format a from import line"""
     parts = []
-    for _, module, name, asname in names:
+    for _, module, name, asname in names:  # noqa
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply amalgamate.py:365]: that module is latter referenced

**Comment by gforsyth** [comment]
I'm not opposed. I think bugbear trips up fewer people because the errors are less common than, say, styling issues.

Two asks:
Can this also be added to the precommit config?
Can you remove the pylint bare except directives since they are no longer bare?

**Comment by jnoortheen** [comment]
> Two asks:
> Can this also be added to the precommit config?

1. do you want to add flake8 as whole? I think we can add it part of push checks. 

> Can you remove the pylint bare except directives since they are no longer bare?
2. will update it.

File: xonsh/completers/python.py [hunk 2]
```diff
@@ -211,11 +211,11 @@ def _safe_eval(expr, ctx):
     try:
         val = xonsh_safe_eval(expr, ctx, ctx, transform=False)
         _ctx = ctx
-    except:  # pylint:disable=bare-except
+    except Exception:  # pylint:disable=bare-except
         try:
             val = xonsh_safe_eval(expr, builtins.__dict__, transform=False)
             _ctx = builtins.__dict__
-        except:  # pylint:disable=bare-except
+        except Exception:  # pylint:disable=bare-except
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/completers/python.py]: removed in other places as well.

**Comment by gforsyth** [comment]
Yeah, let's add flake8.

Then later we can run precommit itself during the style CI runs and so anyone running precommit locally will be guaranteed to pass on CI

File: xonsh/inspectors.py [hunk 3]
```diff
@@ -393,7 +393,7 @@ def pdoc(self, obj, oname=""):
             if init_ds is not None:
                 lines.append(head("Init docstring:"))
                 lines.append(indent(init_ds))
-        elif hasattr(obj, "__call__"):
+        elif hasattr(obj, "__call__"):  # noqa : use callable
```
[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/inspectors.py:396]: here I don't think __call__ and being callable is the same. this wouldn't pass a function and callable would.

**Comment by jnoortheen** [comment]
> Yeah, let's add flake8.
> 
> Then later we can run precommit itself during the style CI runs and so anyone running precommit locally will be guaranteed to pass on CI

added qa as whole to pre-commit hook. Added this as pre-push hook since it will take some time to run.
command to install this hook

```
pre-commit install --hook-type pre-push
```

File: xonsh/inspectors.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniel-shimon) [reply xonsh/inspectors.py:396]: Functions also have `__call__` set.
Callable's implementation checks for `__call__` itself (or checks the c equivalent).
(see `PyCallable_Check` in http://svn.python.org/projects/python/trunk/Objects/object.c)

[the remark this answers is not in the document]
Comment (jnoortheen) [reply xonsh/inspectors.py]: updated

**Comment by jnoortheen** [comment]
what happened to the news-item checking action? it doesn't seem to be running for recent PRs.

**Review by gforsyth — APPROVED** [review]
Changes here look good to me.  @daniel-shimon -- anything else you want to see done here?

As for the news checks, @AstraLuma set them up ages ago and some webhook may not be firing anymore?  
We can probably add a news check with github actions fairly easily.

**Comment by daniel-shimon** [comment]
I think it's good - thanks @jnoortheen 👌

**Pull request merged**